src/ast/nodes/BinaryExpression.ts

```diff
@@ -50,27 +27,28 @@ const operators: {
 	'^': (left: any, right: any) => left ^ right,
 	'&': (left: any, right: any) => left & right,
 	'**': (left: any, right: any) => Math.pow(left, right),
-	in: (left: any, right: any) => left in right,
-	instanceof: (left: any, right: any) => left instanceof right
+	in: (left, right: any) => left in right,
+	instanceof: (left, right: any) => left instanceof right
 };
 
 export default class BinaryExpression extends NodeBase {
 	type: NodeType.BinaryExpression;
 	left: ExpressionNode;
 	right: ExpressionNode;
-	operator: BinaryOperator;
+	operator: keyof typeof binaryOperators;
 
-	getValue(): any {
-		const leftValue = this.left.getValue();
+	getLiteralValueAtPath(path: ObjectPath): LiteralValueOrUnknown {
```

Comment:
I would prefer if we keep this as `getValue` and rather create a wrapper type for `LiteralValue` - sort of thing perhaps. The reason being that it would be nice to look to supporting things like:

```js
var p = {};
if (!p) // treeshake!
  sideEffect();
```

Such a thing could be supported very simply based on this logic if we created a value type for eg "UnknownObject". Such a thing could pass checks like `typeof p === 'object'` as well, etc working with the uncertainty levels.

I'd also hope the value type could ultimately be able to grow towards a full-blown typing system with value unions and intersection types in future. It could well be worth researching common typing implementations for some inspiration on the architecture here too.

Reply:
>  it would be nice to look to supporting things like...

I see the use case of adding additional pseudo types like `UNKNOWN_OBJECT_VALUE` or maybe `UNKNOWN_TRUTHY_VALUE` but I would not like change the method name back to `getValue` as this name seems to imply for many that this is meant as a typing system or similar—and this is NOT the intention of this method (but I am open for other suggestions that better describe the intention outlined below). Most notably because we already have a capable typing system (supporting e.g. union types, see below) where the data flows exactly the opposite way.

The intention of this method is
* return a stringifyable primitive value if possible (suitable for inlining without further checks), and
* provide a way to resolve conditionals

We might extend this in the future but then we can always change the name again. Right now, `getLiteralValueAtPath` is the best I could think of to describe the stringifying intention that should not be broken by future extensions. BTW we already use this inlining in one edge case around "if" statements but I see us doing it more often, e.g. for side-effect free computed properties.

> I'd also hope the value type could ultimately be able to grow towards a full-blown typing system with value unions and intersection types in future

As I said above, we already have a capable typing system that is just somewhat invisible because the types are dynamically refined (or rather generalized) and cannot be queried directly.  Cf. this example of a union type:
```javascript
const x = {
  a: () => {},
  b: 1
};
const y = {
  a: () => {},
  c: 2
};
const z = Math.random() > 0.5 ? x : y;
z.a();
```

This code will be completely tree-shaken even though `z` has a union type and I would rather work extending the existing logic than work against it. In a way, `getLiteralValueAtPath` is my least favourite part of the typing system as it works different from everything else. Maybe I find a better solution in the future. If you want, I can add a comment in `Expression.ts` to outline this intention.

Reply:
I definitely get your point here, and it makes sense now with respect to the code. The test case really helped to illustrate what you mean, and is super impressive! 

The current architecture could be considered as "static single-assignment object path analysis", which does seem like it hits a sweet spot between performance and utility. I'm still hankering after some features of "flow-based execution analysis" here to get deeper optimization on the critical path (which exactly travels in the execution direction like this). 

Perhaps there will be a middle-ground where we can use this static path analysis on the entire codebase (fast), and then have a separate mechanism to run execution flow-analysis on the critical execution path with some cutoffs.

It would be good to discuss this further as this does hit on those architecture questions. My primary concern is just to ensure that we're not driving the project into a local maxima with respect to what optimizations are possible.

Reply:
Not sure if we really need another analysis path this to get information about the critical path. In fact, the current analysis already follows the execution path. If the goal is to identify the critical path, one approach could be to just add another option to the `ExecutionPathOptions` to tell the analysis we are only interested in critical path code and observe this flag in the right places.

My comment about about data flowing the opposite way was probably a bit mis-leading and only about the question whether to
* ask a node if it has e.g. a side-effect and the node defers the request to its children, or
* ask a node to get some kind of "resulting value" and then analyse this "value"

Currently I favour the first approach as it has turned out to be much more powerful while requiring less special handling of certain node types in otherwise unrelated nodes, which is one of my main concerns in keeping the code clean here as not having special handling guarantees safe extensibility.

Reply:
I understand and agree with the motivations here, I'm just weary of us closing doors when we don't have to.

The important side of the critical path analysis that gets missed is variables where their assignments are made separately to their initialization statements. Basically execution ordering.

The sort of use case I have in mind is:

```js
var executed = false;
function execute () {
  executed = true;
}
execute();
if (!executed)
  treeshakeMe();
```

being able to simplify to `export default { hello: 'world' }`. Effectively implementing a value-possibility-range-tracking interpretor with cutoffs for the critical path upon which more advanced optimizations can be done. So seeing `getValue` effectively acting as a limited interpretor here, makes me want to share that logic, but perhaps we could integrate this in other ways.

I can try and get some more common use cases from popular codebases together if it would help here and follow up in the inlining issue.

Reply:
With all respect, you are aware code can and always should be refactored? We are not closing doors as long as we are capable of refactoring things and I will not stop making code work best for the current situation because we "might want something else in the future". Once we actually start implementing this maybe another approach will look more promising? But we make our code less clean and more convoluted now to prepare for this? This is closely related to the http://antipatterns.com/lavaflow.htm pattern IMO.

But in this specific situation, SSA will not happen via getValue any more than it will happen the other way. It is just as easy or difficult to implement. What we actually need for SSA (at least that is what you describe to my understanding) to work is variable versioning.

Ans yes, renaming a method is one of the simplest refactorings my IDE does, takes a second. It's just that I like the name to describe what it does TODAY so that the method is not accidentally misused.

That being said, common use cases from popular codebases are very important to help us decide on how to approach this (though I still think other topics and lower hanging fruits should be approached first; and who knows what the code base will look like then?).

Reply:
That being said, a situation very similar to what you describe might actually be solved by inlining the bind phase into the tree-shaking phase and only bind code that is actually executed, something I already started working on a little in the past but did not go through with yet.

Reply:
> With all respect, you are aware code can and always should be refactored? We are not closing doors as long as we are capable of refactoring things

Certainly, this is really all that's important to me. And I've agreed with your decision, I was only explaining the reason for my suggestion.

Also definitely agreed we should write code to solve the practical problems first and not over-generalize.

To bring this back to my tangible use case - that example I provided is what my CommonJS transform looks like that can handle circular references and exact execution semantics, and I'm hoping to integrate this into a Rollup plugin very soon. So I'm pushing for the optimization opportunities on this because otherwise there are a lot of unnecessary function wrapper calls, and there is no way to do this feature without mutable globals tracking execution ordering.

Reply:
Ah I see, that makes sense. At the moment, there are really several problems that are not quite easy to solve where the biggest issue I see is keeping the performance in sane limits. Is there a way that the assignment to `executed` can happen in the same function scope as the if statement? That would make it a lot easier to add static analysis without breaking the bank.

Also I assume there is some conditional surrounding the assignment to `executed` that is missing in the example?

But maybe this should continue in a separate issue. Improved CJS handling would be a godsend but making it work with tree-shaking is certainly quite important. Maybe we can tweak both the generated code and the tree-shaking algorithm together so that we can find a working solution as soon as possible.